src/convert_to_typescript2.ts

```diff
@@ -0,0 +1,47 @@
+enum EventTypes {
+    'EAT',
+    'SLEEP',
+    'WORK'
+}
+
+type Eat = {
+    food: string
+    type: EventTypes.EAT
+}
+
+type Sleep = {
+    dream: string
+    duration: number
+    type: EventTypes.SLEEP
+}
+
+type Work = {
+    profession: string
+    slack: string
+    type: EventTypes.WORK
+}
+
+type Play = {
+    type: 'PLAY'
+}
+
+type Events = Eat | Sleep | Work
+
+function assertUnreachable(x: never): never {
+    throw new Error("Didn't expect to get here");
+}
+
+//TODO: Couldn't get exhaustive pattern matching working?
```

Comment:
the trick here is to use string for `event.type`, e.g. 
```
type Eat = { type: 'Eat', ... }
```

Reply:
I've changed to string as you suggested, this code does work if you add `Play` to the union. But I get a false negative when the union is just `Eat | Sleep | Work`. The compiler error I get in this case is:

```
TypeScript error in /Users/jim/code/callery/src/convert_to_typescript2.ts(34,36):
Argument of type 'any' is not assignable to parameter of type 'never'.  TS2345

    32 |         case 'SLEEP': return `You will dream about ${event.dream} for ${event.duration} hours`;
    33 |         case 'WORK': return `A job in ${event} is tough but you can always ${event.slack}!`;
  > 34 |         default: assertUnreachable(event.type)
       |                                    ^
    35 |     }
    36 | }
    37 | console.log(handleEvent({ type: 'EAT', food: 'Pizza' }));

```

Maybe we can look at this as part of our review?